File: sdk/python/secure_script.py
"""
安全脚本模块
提供加密脚本的获取、解密和执行功能

安全特性：
- 支持 HTTPS 和证书固定（继承自 LicenseClient）
- AES-GCM 加密
- RSA 签名验证
- HKDF 密钥派生

使用示例：
    from license_client import LicenseClient
    from secure_script import SecureScriptManager

    client = LicenseClient(
        server_url="https://192.168.1.100:8080",
        app_key="your_app_key",
        cert_fingerprint="SHA256:AB:CD:EF:..."  # 证书固定
    )

    # 创建安全脚本管理器
    script_manager = SecureScriptManager(client, app_secret="your_app_secret")

    # 获取脚本版本列表
    versions = script_manager.get_script_versions()

    # 获取并解密脚本
    script = script_manager.fetch_script("script_id")

    # 执行脚本
    result = script_manager.execute_script("script_id", executor=my_executor)
"""

import base64
import hashlib
import json
import threading
import time
from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Optional

try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.hkdf import HKDF
    from cryptography.hazmat.backends import default_backend
    from cryptography.hazmat.primitives.asymmetric import padding
    from cryptography.hazmat.primitives import serialization
except ImportError:
    raise ImportError("请安装 cryptography 库: pip install cryptography")

try:
    import requests
except ImportError:
    raise ImportError("请安装 requests 库: pip install requests")


@dataclass
class CachedScript:
    """缓存的脚本"""
    script_id: str
    version: str
    content: bytes
    content_hash: str
    fetched_at: float
    expires_at: float


@dataclass
class EncryptedScriptPackage:
    """加密脚本包"""
    script_id: str
    version: str
    script_type: str
    entry_point: str
    encrypted_content: str
    content_hash: str
    key_hint: str
    signature: str
    expires_at: int
    timeout: int
    memory_limit: int
    parameters: str
    execute_once: bool


@dataclass
class ScriptVersionInfo:
    """脚本版本信息"""
    script_id: str
    name: str
    version: str
    content_hash: str
    updated_at: int


class SecureScriptError(Exception):
    """安全脚本错误"""
    pass


class SecureScriptManager:
    """安全脚本管理器"""

    def __init__(
        self,
        client,  # LicenseClient 实例
        app_secret: str,
        on_execute: Optional[Callable[[str, str, Optional[Exception]], None]] = None
    ):
        """
        初始化安全脚本管理器

        Args:
            client: LicenseClient 实例
            app_secret: 应用密钥 (用于密钥派生)
            on_execute: 执行回调函数 (script_id, status, error)
        """
        self.client = client
        self.app_secret = app_secret
        self.on_execute = on_execute
        self._script_cache: Dict[str, CachedScript] = {}
        self._lock = threading.Lock()

    def get_script_versions(self) -> List[ScriptVersionInfo]:
        """
        获取脚本版本列表

        Returns:
            脚本版本信息列表
        """
        url = f"{self.client.server_url}/api/client/secure-scripts/versions"
        params = {"app_key": self.client.app_key}

        resp = requests.get(url, params=params, timeout=30)
        result = resp.json()

        if result.get('code') != 0:
            raise SecureScriptError(result.get('message', '获取版本失败'))

        versions = []
        for item in result.get('data', []):
            versions.append(ScriptVersionInfo(
                script_id=item['script_id'],
                name=item['name'],
                version=item['version'],
                content_hash=item['content_hash'],
                updated_at=item['updated_at']
            ))

        return versions

    def fetch_script(self, script_id: str, force: bool = False) -> CachedScript:
        """
        获取加密脚本并解密

        Args:
            script_id: 脚本ID
            force: 是否强制刷新缓存

        Returns:
            解密后的脚本
        """
        # 检查缓存
        if not force:
            with self._lock:
                if script_id in self._script_cache:
                    cached = self._script_cache[script_id]
                    if time.time() < cached.expires_at:
                        return cached

        # 请求服务器
        url = f"{self.client.server_url}/api/client/secure-scripts/fetch"
        data = {
            "app_key": self.client.app_key,
            "machine_id": self.client.machine_id,
            "script_id": script_id
        }

        resp = requests.post(url, json=data, timeout=30)
        result = resp.json()

        if result.get('code') != 0:
            raise SecureScriptError(result.get('message', '获取脚本失败'))

        pkg_data = result.get('data', {})
        pkg = EncryptedScriptPackage(
            script_id=pkg_data['script_id'],
            version=pkg_data['version'],
            script_type=pkg_data['script_type'],
            entry_point=pkg_data['entry_point'],
            encrypted_content=pkg_data['encrypted_content'],
            content_hash=pkg_data['content_hash'],
            key_hint=pkg_data['key_hint'],
            signature=pkg_data['signature'],
            expires_at=pkg_data['expires_at'],
            timeout=pkg_data['timeout'],
            memory_limit=pkg_data['memory_limit'],
            parameters=pkg_data['parameters'],
            execute_once=pkg_data['execute_once']
        )

        # 验证过期时间
        if time.time() > pkg.expires_at:
            raise SecureScriptError("脚本包已过期")

        # 验证签名
        sign_data = f"{pkg.script_id}:{pkg.encrypted_content}:{self.client.machine_id}:{pkg.expires_at}"
        if not self._verify_signature(sign_data.encode(), pkg.signature):
            raise SecureScriptError("签名验证失败")

        # 派生解密密钥
        decrypt_key = self._derive_key(pkg.key_hint)

        # 解密内容
        content = self._decrypt_content(pkg.encrypted_content, decrypt_key)

        # 验证哈希
        content_hash = hashlib.sha256(content).hexdigest()
        if content_hash != pkg.content_hash:
            raise SecureScriptError("内容哈希不匹配")

        # 缓存
        cached = CachedScript(
            script_id=pkg.script_id,
            version=pkg.version,
            content=content,
            content_hash=pkg.content_hash,
            fetched_at=time.time(),
            expires_at=pkg.expires_at
        )

        with self._lock:
            self._script_cache[script_id] = cached

        return cached

    def execute_script(
        self,
        script_id: str,
        args: Optional[Dict[str, Any]] = None,
        executor: Optional[Callable[[bytes, Dict[str, Any]], str]] = None
    ) -> str:
        """
        执行脚本

        Args:
            script_id: 脚本ID
            args: 执行参数
            executor: 执行函数，接收 (content, args)，返回结果字符串

        Returns:
            执行结果
        """
        if executor is None:
            # 默认使用 exec 执行 Python 脚本
            def default_executor(content: bytes, args: Dict[str, Any]) -> str:
                local_vars = {"args": args or {}, "result": None}
                exec(content.decode('utf-8'), {}, local_vars)
                return str(local_vars.get('result', ''))
            executor = default_executor

        # 获取脚本
        try:
            script = self.fetch_script(script_id)
        except Exception as e:
            self._report_execution(script_id, "", "failed", "", str(e), 0)
            if self.on_execute:
                self.on_execute(script_id, "failed", e)
            raise

        # 上报开始执行
        self._report_execution(script_id, "", "executing", "", "", 0)
        if self.on_execute:
            self.on_execute(script_id, "executing", None)

        # 执行
        start_time = time.time()
        try:
            result = executor(script.content, args or {})
            duration = int((time.time() - start_time) * 1000)

            # 上报成功
            self._report_execution(script_id, "", "success", result, "", duration)
            if self.on_execute:
                self.on_execute(script_id, "success", None)

            # 清除缓存 (安全考虑)
            with self._lock:
                self._script_cache.pop(script_id, None)

            return result

        except Exception as e:
            duration = int((time.time() - start_time) * 1000)

            # 上报失败
            self._report_execution(script_id, "", "failed", "", str(e), duration)
            if self.on_execute:
                self.on_execute(script_id, "failed", e)

            # 清除缓存
            with self._lock:
                self._script_cache.pop(script_id, None)

            raise

    def clear_cache(self):
        """清除脚本缓存"""
        with self._lock:
            self._script_cache.clear()

    def get_cached_script(self, script_id: str) -> Optional[CachedScript]:
        """获取缓存的脚本 (不请求服务器)"""
        with self._lock:
            return self._script_cache.get(script_id)

    # 内部方法

    def _derive_key(self, key_hint: str) -> bytes:
        """派生解密密钥"""
        master_key = self.app_secret.encode()
        salt = self.client.machine_id.encode()
        info = key_hint.encode()

        hkdf = HKDF(
            algorithm=hashes.SHA256(),
            length=32,
            salt=salt,
            info=info,
            backend=default_backend()
        )

        return hkdf.derive(master_key)

    def _decrypt_content(self, encrypted_base64: str, key: bytes) -> bytes:
        """解密内容"""
        ciphertext = base64.b64decode(encrypted_base64)

        # AES-GCM: nonce (12 bytes) + ciphertext + tag (16 bytes)
        nonce_size = 12
        if len(ciphertext) < nonce_size:
            raise SecureScriptError("密文太短")

        nonce = ciphertext[:nonce_size]
        ciphertext = ciphertext[nonce_size:]

        aesgcm = AESGCM(key)
        return aesgcm.decrypt(nonce, ciphertext, None)

    def _verify_signature(self, data: bytes, signature_base64: str) -> bool:
        """验证签名"""
        if not hasattr(self.client, 'public_key') or self.client.public_key is None:
            return True  # 如果没有公钥，跳过验证

        try:
            signature = base64.b64decode(signature_base64)
            self.client.public_key.verify(
                signature,
                data,
                padding.PKCS1v15(),
                hashes.SHA256()
            )
            return True
        except Exception:
            return False

    def _report_execution(
        self,
        script_id: str,
        delivery_id: str,
        status: str,
        result: str,
        error_msg: str,
        duration: int
    ):
        """上报执行状态"""
        try:
            url = f"{self.client.server_url}/api/client/secure-scripts/report"
            data = {
                "app_key": self.client.app_key,
                "machine_id": self.client.machine_id,
                "script_id": script_id,
                "status": status
            }
            if delivery_id:
                data["delivery_id"] = delivery_id
            if result:
                data["result"] = result
            if error_msg:
                data["error_message"] = error_msg
            if duration > 0:
                data["duration"] = duration

            # 异步上报
            threading.Thread(
                target=lambda: requests.post(url, json=data, timeout=10),
                daemon=True
            ).start()
        except Exception:
            pass


# 便捷函数

def fetch_and_execute(
    client,
    app_secret: str,
    script_id: str,
    args: Optional[Dict[str, Any]] = None,
    executor: Optional[Callable[[bytes, Dict[str, Any]], str]] = None
) -> str:
    """
    获取并执行脚本 (便捷函数)

    Args:
        client: LicenseClient 实例
        app_secret: 应用密钥
        script_id: 脚本ID
        args: 执行参数
        executor: 执行函数

    Returns:
        执行结果
    """
    manager = SecureScriptManager(client, app_secret)
    return manager.execute_script(script_id, args, executor)

File: sdk/go/secure_script.go
package license

import (
	"crypto"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
	"time"

	"golang.org/x/crypto/hkdf"
)

// SecureScriptManager 安全脚本管理器
type SecureScriptManager struct {
	client       *Client
	appSecret    string
	publicKey    *rsa.PublicKey
	scriptCache  map[string]*CachedScript
	mu           sync.RWMutex
	onExecute    func(scriptID string, status string, err error)
}

// CachedScript 缓存的脚本
type CachedScript struct {
	ScriptID    string
	Version     string
	Content     []byte // 解密后的内容
	ContentHash string
	FetchedAt   time.Time
	ExpiresAt   time.Time
}

// EncryptedScriptPackage 加密脚本包
type EncryptedScriptPackage struct {
	ScriptID         string `json:"script_id"`
	Version          string `json:"version"`
	ScriptType       string `json:"script_type"`
	EntryPoint       string `json:"entry_point"`
	EncryptedContent string `json:"encrypted_content"`
	ContentHash      string `json:"content_hash"`
	KeyHint          string `json:"key_hint"`
	Signature        string `json:"signature"`
	ExpiresAt        int64  `json:"expires_at"`
	Timeout          int    `json:"timeout"`
	MemoryLimit      int    `json:"memory_limit"`
	Parameters       string `json:"parameters"`
	ExecuteOnce      bool   `json:"execute_once"`
}

// ScriptVersionInfo 脚本版本信息
type ScriptVersionInfo struct {
	ScriptID    string `json:"script_id"`
	Name        string `json:"name"`
	Version     string `json:"version"`
	ContentHash string `json:"content_hash"`
	UpdatedAt   int64  `json:"updated_at"`
}

// SecureScriptOption 配置选项
type SecureScriptOption func(*SecureScriptManager)

// WithAppSecret 设置应用密钥 (用于密钥派生)
func WithAppSecret(secret string) SecureScriptOption {
	return func(m *SecureScriptManager) {
		m.appSecret = secret
	}
}

// WithExecuteCallback 设置执行回调
func WithExecuteCallback(callback func(scriptID string, status string, err error)) SecureScriptOption {
	return func(m *SecureScriptManager) {
		m.onExecute = callback
	}
}

// WithPublicKey 设置公钥 (用于签名验证)
func WithPublicKey(publicKey *rsa.PublicKey) SecureScriptOption {
	return func(m *SecureScriptManager) {
		m.publicKey = publicKey
	}
}

// NewSecureScriptManager 创建安全脚本管理器
func NewSecureScriptManager(client *Client, opts ...SecureScriptOption) *SecureScriptManager {
	m := &SecureScriptManager{
		client:      client,
		scriptCache: make(map[string]*CachedScript),
	}

	for _, opt := range opts {
		opt(m)
	}

	return m
}

// GetScriptVersions 获取脚本版本列表
func (m *SecureScriptManager) GetScriptVersions() ([]ScriptVersionInfo, error) {
	url := fmt.Sprintf("%s/api/client/secure-scripts/versions?app_key=%s",
		m.client.GetServerURL(), m.client.GetAppKey())

	resp, err := m.client.GetHTTPClient().Get(url)
	if err != nil {
		return nil, fmt.Errorf("请求失败: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("读取响应失败: %w", err)
	}

	var result struct {
		Code    int                 `json:"code"`
		Message string              `json:"message"`
		Data    []ScriptVersionInfo `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("解析响应失败: %w", err)
	}

	if result.Code != 0 {
		return nil, fmt.Errorf("API错误: %s", result.Message)
	}

	return result.Data, nil
}

// FetchScript 获取加密脚本并解密
func (m *SecureScriptManager) FetchScript(scriptID string) (*CachedScript, error) {
	// 检查缓存
	m.mu.RLock()
	if cached, ok := m.scriptCache[scriptID]; ok {
		if time.Now().Before(cached.ExpiresAt) {
			m.mu.RUnlock()
			return cached, nil
		}
	}
	m.mu.RUnlock()

	// 请求服务器
	reqBody := map[string]string{
		"app_key":    m.client.GetAppKey(),
		"machine_id": m.client.GetMachineID(),
		"script_id":  scriptID,
	}

	data, err := m.client.request("POST", "/secure-scripts/fetch", reqBody)
	if err != nil {
		return nil, fmt.Errorf("获取脚本失败: %w", err)
	}

	// 将 map 转换为 JSON 再解析
	dataBytes, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("序列化数据失败: %w", err)
	}

	var pkg EncryptedScriptPackage
	if err := json.Unmarshal(dataBytes, &pkg); err != nil {
		return nil, fmt.Errorf("解析脚本包失败: %w", err)
	}

	// 验证过期时间
	if time.Now().Unix() > pkg.ExpiresAt {
		return nil, errors.New("脚本包已过期")
	}

	// 验证签名
	signData := fmt.Sprintf("%s:%s:%s:%d", pkg.ScriptID, pkg.EncryptedContent, m.client.GetMachineID(), pkg.ExpiresAt)
	if err := m.verifySignature([]byte(signData), pkg.Signature); err != nil {
		return nil, fmt.Errorf("签名验证失败: %w", err)
	}

	// 派生解密密钥
	decryptKey, err := m.deriveKey(pkg.KeyHint)
	if err != nil {
		return nil, fmt.Errorf("派生密钥失败: %w", err)
	}

	// 解密内容
	content, err := m.decryptContent(pkg.EncryptedContent, decryptKey)
	if err != nil {
		return nil, fmt.Errorf("解密失败: %w", err)
	}

	// 验证哈希
	contentHash := sha256Hash(content)
	if contentHash != pkg.ContentHash {
		return nil, errors.New("内容哈希不匹配")
	}

	// 缓存
	cached := &CachedScript{
		ScriptID:    pkg.ScriptID,
		Version:     pkg.Version,
		Content:     content,
		ContentHash: pkg.ContentHash,
		FetchedAt:   time.Now(),
		ExpiresAt:   time.Unix(pkg.ExpiresAt, 0),
	}

	m.mu.Lock()
	m.scriptCache[scriptID] = cached
	m.mu.Unlock()

	return cached, nil
}

// ExecuteScript 执行脚本 (需要外部执行器)
// executor: 执行函数，接收脚本内容和参数，返回结果和错误
func (m *SecureScriptManager) ExecuteScript(
	scriptID string,
	args map[string]interface{},
	executor func(content []byte, args map[string]interface{}) (string, error),
) (string, error) {
	// 获取脚本
	script, err := m.FetchScript(scriptID)
	if err != nil {
		m.reportExecution(scriptID, "", "failed", "", err.Error(), 0)
		if m.onExecute != nil {
			m.onExecute(scriptID, "failed", err)
		}
		return "", err
	}

	// 上报开始执行
	m.reportExecution(scriptID, "", "executing", "", "", 0)
	if m.onExecute != nil {
		m.onExecute(scriptID, "executing", nil)
	}

	// 执行
	startTime := time.Now()
	result, execErr := executor(script.Content, args)
	duration := int(time.Since(startTime).Milliseconds())

	// 上报结果
	status := "success"
	errMsg := ""
	if execErr != nil {
		status = "failed"
		errMsg = execErr.Error()
	}

	m.reportExecution(scriptID, "", status, result, errMsg, duration)
	if m.onExecute != nil {
		m.onExecute(scriptID, status, execErr)
	}

	// 执行后清除缓存 (安全考虑)
	m.mu.Lock()
	delete(m.scriptCache, scriptID)
	m.mu.Unlock()

	return result, execErr
}

// ClearCache 清除脚本缓存
func (m *SecureScriptManager) ClearCache() {
	m.mu.Lock()
	m.scriptCache = make(map[string]*CachedScript)
	m.mu.Unlock()
}

// GetCachedScript 获取缓存的脚本 (不请求服务器)
func (m *SecureScriptManager) GetCachedScript(scriptID string) *CachedScript {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.scriptCache[scriptID]
}

// 内部方法

func (m *SecureScriptManager) deriveKey(keyHint string) ([]byte, error) {
	if m.appSecret == "" {
		return nil, errors.New("未设置应用密钥")
	}

	masterKey := []byte(m.appSecret)
	salt := m.client.GetMachineID()
	info := keyHint

	hash := sha256.New
	hkdfReader := hkdf.New(hash, masterKey, []byte(salt), []byte(info))

	derivedKey := make([]byte, 32)
	if _, err := io.ReadFull(hkdfReader, derivedKey); err != nil {
		return nil, err
	}

	return derivedKey, nil
}

func (m *SecureScriptManager) decryptContent(encryptedBase64 string, key []byte) ([]byte, error) {
	ciphertext, err := base64.StdEncoding.DecodeString(encryptedBase64)
	if err != nil {
		return nil, err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	nonceSize := gcm.NonceSize()
	if len(ciphertext) < nonceSize {
		return nil, errors.New("密文太短")
	}

	nonce, ciphertext := ciphertext[:nonceSize], ciphertext[nonceSize:]
	return gcm.Open(nil, nonce, ciphertext, nil)
}

func (m *SecureScriptManager) verifySignature(data []byte, signatureBase64 string) error {
	if m.publicKey == nil {
		return nil // 如果没有设置公钥，跳过验证
	}

	signature, err := base64.StdEncoding.DecodeString(signatureBase64)
	if err != nil {
		return err
	}

	hashed := sha256.Sum256(data)
	return rsa.VerifyPKCS1v15(m.publicKey, crypto.SHA256, hashed[:], signature)
}

func (m *SecureScriptManager) reportExecution(scriptID, deliveryID, status, result, errorMsg string, duration int) {
	reqBody := map[string]interface{}{
		"app_key":    m.client.GetAppKey(),
		"machine_id": m.client.GetMachineID(),
		"script_id":  scriptID,
		"status":     status,
	}
	if deliveryID != "" {
		reqBody["delivery_id"] = deliveryID
	}
	if result != "" {
		reqBody["result"] = result
	}
	if errorMsg != "" {
		reqBody["error_message"] = errorMsg
	}
	if duration > 0 {
		reqBody["duration"] = duration
	}

	// 异步上报
	go m.client.request("POST", "/secure-scripts/report", reqBody)
}

func sha256Hash(data []byte) string {
	hash := sha256.Sum256(data)
	return hex.EncodeToString(hash[:])
}
